core/packages/contracts/contracts/test/foundry/BeefyClient.t.sol

```diff
@@ -0,0 +1,227 @@
+        randaoCommitExpiration = 8;
+        difficulty = 377;
+
+        beefyClient = new BeefyClient(randaoCommitDelay, randaoCommitExpiration);
+
+        // Allocate for input variables
+        string[] memory inputs = new string[](10);
+        inputs[0] = "test/ffiWrapper.ts";
+        // Always add type of command as first arguments
+        inputs[1] = "GenerateInitialSet";
+        // generate initial fixture data with ffi
+        (blockNumber, setId, setSize, bitSetArray, commitHash, payload) = abi.decode(
+            vm.ffi(inputs),
+            (uint32, uint32, uint32, uint256[], bytes32, BeefyClient.Payload)
+        );
+        bitfield = Bitfield.createBitfield(bitSetArray, setSize);
+
+        // To avoid another round of ffi in multiple tests
+        // except for the initial merkle root and proof for validators
+        // we also precalculate finalValidatorProofs and cached here
+        finalBitfield = Bitfield.randomNBitsWithPriorCheck(
+            difficulty,
+            bitfield,
+            minimumSignatureThreshold(setSize),
+            setSize
+        );
+
+        inputs[1] = "GenerateProofs";
+        // First add length of finalBitField and then each item as input
+        inputs[2] = Strings.toString(finalBitfield.length);
+        for (uint i = 0; i < finalBitfield.length; i++) {
+            inputs[i + 3] = Strings.toString(finalBitfield[i]);
+        }
+        BeefyClient.ValidatorProof[] memory _proofs;
+        (root, validatorProof, _proofs, mmrLeafProofs, mmrLeaf) = abi.decode(
+            vm.ffi(inputs),
+            (
+                bytes32,
+                BeefyClient.ValidatorProof,
+                BeefyClient.ValidatorProof[],
+                bytes32[],
+                BeefyClient.MMRLeaf
+            )
+        );
+        // Cache finalValidatorProofs to storage in order to reuse in submitFinal later
+        for (uint i = 0; i < _proofs.length; i++) {
+            finalValidatorProofs.push(_proofs[i]);
+        }
+        console.log(
+            "current validator's merkle root is: %s",
+            Strings.toHexString(uint256(root), 32)
+        );
+    }
+
+    function testSubmit() public {
+        currentSetId = setId;
+        nextSetId = setId + 1;
+        BeefyClient.ValidatorSet memory vset = BeefyClient.ValidatorSet(
+            currentSetId,
+            setSize,
+            root
+        );
+        BeefyClient.ValidatorSet memory nextvset = BeefyClient.ValidatorSet(
+            nextSetId,
+            setSize,
+            root
+        );
+        beefyClient.initialize(0, vset, nextvset);
+
+        beefyClient.submitInitial(commitHash, bitfield, validatorProof);
+
+        // mine 3 blocks
+        cheats.roll(block.number + randaoCommitDelay);
+        // set difficulty as PrevRandao
+        cheats.difficulty(difficulty);
+
+        beefyClient.commitPrevRandao(commitHash);
+
+        beefyClient.createFinalBitfield(commitHash, bitfield);
+
```

Comment:
This line seems like a noop? Perhaps we should remove it entirely?

Reply:
yes, useless now